evm/src/contracts/WormholeCctp.sol

```diff
@@ -0,0 +1,268 @@
+// SPDX-License-Identifier: Apache 2
+
+pragma solidity ^0.8.22;
+
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import {SafeERC20} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+import {IWormhole} from "src/interfaces/IWormhole.sol";
+import {IMessageTransmitter} from "src/interfaces/IMessageTransmitter.sol";
+import {ITokenMessenger} from "src/interfaces/ITokenMessenger.sol";
+import {ITokenMinter} from "src/interfaces/ITokenMinter.sol";
+
+import {Utils} from "src/libraries/Utils.sol";
+import {WormholeCctpMessages} from "src/libraries/WormholeCctpMessages.sol";
+
+abstract contract WormholeCctp {
+    using Utils for address;
+    using SafeERC20 for IERC20;
+    using WormholeCctpMessages for *;
+
+    error InvalidVaa();
+    error CallerNotMintRecipient(bytes32, bytes32);
+    error CctpVaaMismatch(uint32, uint32, uint64);
+
+    IWormhole immutable _wormhole;
+    uint16 immutable _chainId;
+    uint8 immutable _messageFinality;
+
+    IMessageTransmitter immutable _messageTransmitter;
+    ITokenMessenger immutable _tokenMessenger;
+    ITokenMinter immutable _tokenMinter;
+    uint32 immutable _localCctpDomain;
+
+    constructor(address wormhole, uint8 messageFinality, address cctpTokenMessenger) {
+        _wormhole = IWormhole(wormhole);
+        _chainId = _wormhole.chainId();
+        _messageFinality = messageFinality;
+
+        _tokenMessenger = ITokenMessenger(cctpTokenMessenger);
+        _messageTransmitter = _tokenMessenger.localMessageTransmitter();
+        _tokenMinter = _tokenMessenger.localMinter();
+        _localCctpDomain = _messageTransmitter.localDomain();
+    }
+
+    function setTokenMessengerApproval(address token, uint256 amount) internal {
+        IERC20(token).forceApprove(address(_tokenMessenger), amount);
```

Comment:
Would it not be more gas efficient to simply use `approve`? The `forceApprove` method uses 2 underlying calls to `approve` and it's only required for tokens that don't allow approvals from non-zero starting values. USDC isn't one of those weird tokens.

Reply:
Good catch, changed it here: https://github.com/wormhole-foundation/wormhole-circle-integration/pull/48/commits/da6a7e324997c6ed24f6e1d1f6dd569e28405fc7

Reply:
Also doubt future Circle tokens would be weird